socketcan/libcanard/src/socketcan.c

```diff
@@ -162,43 +179,104 @@ int16_t socketcanPop(const SocketCANFD       fd,
+        const size_t control_size = sizeof(struct cmsghdr) + sizeof(struct timeval);
+        uint8_t*     control      = (uint8_t*) malloc(control_size);  // TODO: try to get rig of malloc!
+        if (control == NULL)
         {
-            return getNegatedErrno();
+            return -ENOMEM;
         }
+        (void) memset(control, 0, control_size);
 
-        // We use the CAN FD struct regardless of whether the CAN FD socket option is set.
-        // Per the user manual, this is acceptable because they are binary compatible.
-        struct canfd_frame cfd;
-        const ssize_t      read_size = read(fd, &cfd, sizeof(cfd));
-        if (read_size < 0)
+        // Initialize the message header used by recvmsg.
+        struct msghdr msg  = {0};           // Message header struct.
+        msg.msg_iov        = &iov;          // Scatter/gather array.
+        msg.msg_iovlen     = 1;             // Number of elements in the scatter/gather array.
+        msg.msg_control    = control;       // Ancillary data.
+        msg.msg_controllen = control_size;  // Ancillary data buffer length.
+
+        // Non-blocking receive messages from the socket and validate.
+        const ssize_t read_size = recvmsg(fd, &msg, MSG_DONTWAIT);
+        if (read_size <= 0)
         {
-            return getNegatedErrno();
+            const int neg_errno = getNegatedErrno();
+
+            // Return either the negated error code, or zero in case the socket is marked nonblocking and the receive
+            // operation would block. POSIX.1-2001 allows either EWOULDBLOCK or EAGAIN to be returned for this case,
+            // and does not require these constants to have the same value. Thus, we check for both possibilities.
+            free(control);
+            return (read_size < 0 && (neg_errno == -EWOULDBLOCK || neg_errno == -EAGAIN)) ? 0 : neg_errno;
         }
         if ((read_size != CAN_MTU) && (read_size != CANFD_MTU))
         {
+            free(control);
             return -EIO;
         }
-        if (cfd.len > payload_buffer_size)
+        if (sockcan_frame.len > payload_buffer_size)
         {
+            free(control);
             return -EFBIG;
         }
 
-        const bool valid = ((cfd.can_id & CAN_EFF_FLAG) != 0) &&  // Extended frame
-                           ((cfd.can_id & CAN_RTR_FLAG) == 0) &&  // Not RTR frame
-                           ((cfd.can_id & CAN_ERR_FLAG) == 0);    // Not error frame
+        const bool valid = ((sockcan_frame.can_id & CAN_EFF_FLAG) != 0) &&  // Extended frame
+                           ((sockcan_frame.can_id & CAN_ERR_FLAG) == 0) &&  // Not RTR frame
+                           ((sockcan_frame.can_id & CAN_RTR_FLAG) == 0);    // Not error frame
         if (!valid)
         {
+            free(control);
             return 0;  // Not an extended data frame -- drop silently and return early.
         }
 
+        const bool loopback_frame = (msg.msg_flags & (int) MSG_CONFIRM) != 0;  // NOLINT
```

Comment:
Bitwise operations on negatives are implementation-defined. Some static analyzers complain when you perform bitwise operations on signed integers for this reason.

```suggestion
        const bool loopback_frame = ((uint32_t) msg.msg_flags & (uint32_t) MSG_CONFIRM) != 0;  // NOLINT
```

Reply:
Gotcha.